### PR TITLE
Generate self-signed certificates with dotnet, powershell, openssl

### PR DESCRIPTION
**Pull request by plooploops**
## Summary

Generate self-signed certificates with `dotnet dev-certs`, `openssl`, and `PowerShell`.  Provide testing samples based off https://github.com/dotnet/dotnet-docker/blob/master/samples/run-aspnetcore-https-development.md

Fixes #Issue_Number (if available)

### CODE REVIEW

**Comment by plooploops** [comment]
FYI: @BillWagner @blowdart @Rick-Anderson

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+
+Depending on the host os, the certificate will need to be trusted. On a `Linux` host, 'trusting' the certificate is different and distro dependent.
+
+For the purposes of this guide, here's an example in `Windows` using `PowerShell`:
+
+```powershell
+Import-Certificate -FilePath $certFilePath -CertStoreLocation 'Cert:\LocalMachine\Root'
+```
+
+For `.netcore 3.1`, run the following command in `wsl`:
```
Comment (BillWagner) [inline docs/core/additional-tools/self-signed-certificates-guide.md:298]: code fence:

```suggestion
For .NET Core 3.1, run the following command in WSL:
```

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+
+```bash
+docker run --rm -it -p 8000:80 -p 8001:443 -e ASPNETCORE_URLS="https://+;http://+" -e ASPNETCORE_HTTPS_PORT=8001 -e ASPNETCORE_ENVIRONMENT=Development -e ASPNETCORE_Kestrel__Certificates__Default__Password="password" -e ASPNETCORE_Kestrel__Certificates__Default__Path=/https/contoso.com.pfx -v /c/path/to/certs/:/https/ mcr.microsoft.com/dotnet/samples:aspnetapp
+```
+
+Starting with `.net 5`, Kestrel can take the `.crt` and PEM-encoded `.key` files. You can run the sample with the following command for `.net 5`:
```
Comment (BillWagner) [inline docs/core/additional-tools/self-signed-certificates-guide.md:304]:
```suggestion
Starting with .NET 5, Kestrel can take the `.crt` and PEM-encoded `.key` files. You can run the sample with the following command for .NET 5:
```

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+
+```bash
+docker run --rm -it -p 8000:80 -p 8001:443 -e ASPNETCORE_URLS="https://+;http://+" -e ASPNETCORE_HTTPS_PORT=8001 -e ASPNETCORE_ENVIRONMENT=Development -e ASPNETCORE_Kestrel__Certificates__Default__Path=/https/contoso.com.crt -e ASPNETCORE_Kestrel__Certificates__Default__KeyPath=/https/contoso.com.key -v /c/path/to/certs:/https/ mcr.microsoft.com/dotnet/samples:aspnetapp
+```
+
+> Note that in WSL, the volume mount path may change depending on the configuration.
```
Comment (BillWagner) [inline docs/core/additional-tools/self-signed-certificates-guide.md:310]: Note styling:

```suggestion
> [!NOTE]
> In WSL, the volume mount path may change depending on the configuration.
```

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+
+For `.netcore 3.1` in Windows, run the following command in `Powershell`:
```
Comment (BillWagner) [inline docs/core/additional-tools/self-signed-certificates-guide.md:312]: code fence:

```suggestion
For .NET Core 3.1 in Windows, run the following command in Powershell:
```

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+
+```powershell
+docker run --rm -it -p 8000:80 -p 8001:443 -e ASPNETCORE_URLS="https://+;http://+" -e ASPNETCORE_HTTPS_PORT=8001 -e ASPNETCORE_ENVIRONMENT=Development -e ASPNETCORE_Kestrel__Certificates__Default__Password="password" -e ASPNETCORE_Kestrel__Certificates__Default__Path=c:\https\contoso.com.pfx -v c:\certs:C:\https aspnetapp:my-sample
+```
+
+For `.net 5` in Windows, run the following command in `PowerShell`:
```
Comment (BillWagner) [inline docs/core/additional-tools/self-signed-certificates-guide.md:318]: code fence:

```suggestion
For .NET 5 in Windows, run the following command in PowerShell:
```

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+
+```powershell
+docker run --rm -it -p 8000:80 -p 8001:443 -e ASPNETCORE_URLS="https://+;http://+" -e ASPNETCORE_HTTPS_PORT=8001 -e ASPNETCORE_ENVIRONMENT=Development -e ASPNETCORE_Kestrel__Certificates__Default__Path=c:\https\contoso.com.crt -e ASPNETCORE_Kestrel__Certificates__Default__KeyPath=c:\https\contoso.com.key -v c:\certs:C:\https aspnetapp:my-sample
+```
+
+Once the application is up, navigate to contoso.com:8001 in a browser.
+
+> Be sure that the host entries are updated for `contoso.com` to answer on  the appropriate ip address (for example 127.0.0.1). If the certificate isn't recognized, make sure that the certificate that is loaded with the container is also trusted on the host, and that there's appropriate SAN / DNS entries for `contoso.com`.
```
Comment (BillWagner) [inline docs/core/additional-tools/self-signed-certificates-guide.md:326]: I'd remove the indenting here:

```suggestion
Be sure that the host entries are updated for `contoso.com` to answer on  the appropriate ip address (for example 127.0.0.1). If the certificate isn't recognized, make sure that the certificate that is loaded with the container is also trusted on the host, and that there's appropriate SAN / DNS entries for `contoso.com`.
```

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+
+If you're using the container built earlier for `Windows`, the run command would look like the following:
+
+```console
+docker run --rm -it -p 8000:80 -p 8001:443 -e ASPNETCORE_URLS="https://+;http://+" -e ASPNETCORE_HTTPS_PORT=8001 -e ASPNETCORE_ENVIRONMENT=Development -e ASPNETCORE_Kestrel__Certificates__Default__Password="password" -e ASPNETCORE_Kestrel__Certificates__Default__Path=c:\https\contoso.com.pfx -v c:\certs:C:\https aspnetapp:my-sample
+```
+
+Once the application is up, navigate to contoso.com:8001 in a browser.
+
+> Be sure that the host entries are updated for `contoso.com` to answer on  the appropriate ip address (for example 127.0.0.1). If the certificate isn't recognized, make sure that the certificate that is loaded with the container is also trusted on the host, and that there's appropriate SAN / DNS entries for `contoso.com`.
```
Comment (BillWagner) [inline docs/core/additional-tools/self-signed-certificates-guide.md:230]: I'd remove the indenting here:

```suggestion
Be sure that the host entries are updated for `contoso.com` to answer on  the appropriate ip address (for example 127.0.0.1). If the certificate isn't recognized, make sure that the certificate that is loaded with the container is also trusted on the host, and that there's appropriate SAN / DNS entries for `contoso.com`.
```

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+
+This sample requires [Docker 17.06](https://docs.docker.com/release-notes/docker-ce) or later of the [Docker client](https://www.docker.com/products/docker).
+
+## Prepare Sample App
+
+You'll need to prepare the sample app depending on which runtime you'd like to use for testing.
+
+For this guide, you'll be using a [sample app](https://hub.docker.com/_/microsoft-dotnet-samples) and make changes where appropriate.
+
+### Prepare .netcore 3.1 Sample App
```
Comment (BillWagner) [inline docs/core/additional-tools/self-signed-certificates-guide.md:32]: use sentence casing and product name:

```suggestion
### Prepare .NET Core 3.1 sample app
```

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+cd .\dotnet-docker\samples\aspnetapp
+```
+
+Build the container for testing locally.
+
+```console
+docker build -t aspnetapp:my-sample -f Dockerfile .
+```
+
+### Prepare .net 5 Sample App
```
Comment (BillWagner) [inline docs/core/additional-tools/self-signed-certificates-guide.md:94]: capitalize .NET, sentence case heading:: 
```suggestion
### Prepare .NET 5 sample app
```

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+
+For `dotnet dev-certs`, be sure to have the appropriate version of `dotnet` installed:
+
+* [Install dotnet on Windows](../install/windows.md)
+* [Install dotnet on Linux](../install/linux.md)
+* [Install dotnet on macOS](../install/macos.md)
+
+This sample requires [Docker 17.06](https://docs.docker.com/release-notes/docker-ce) or later of the [Docker client](https://www.docker.com/products/docker).
+
+## Prepare Sample App
```
Comment (BillWagner) [inline docs/core/additional-tools/self-signed-certificates-guide.md:26]: I didn't mark this on every heading, but all headings should be sentence case, not title case:

```suggestion
## Prepare sample app
```

**Review by BillWagner — CHANGES_REQUESTED** [review]
Overall, this looks great @plooploops 

I had a number of style comments, and left suggestions for almost all of them. (I didn't add suggestions for all of the sentence case headings). With some of the suggestions, I wasn't sure which style would be best, so I added extra notes to consider.

Once these are incorporated, this is ready to merge.

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 11]
```diff
@@ -0,0 +1,334 @@
+---
+title: Generate Self-Signed Certificates Overview
+description: An overview of the Microsoft dotnet dev-certs tool that adds functionality for .NET Core and ASP.NET Core projects, and other options for using self-signed certificates.
+author: angee
+ms.date: 11/19/2020
+---
+
+# Generating Self-Signed Certificates
```
[the remark this answers is not in the document]
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+
+When using self-signed certificates, there's different ways to create and use them for development and testing scenarios.  In this guide, you'll cover using self-signed certificates with `dotnet dev-certs`, and other options like `PowerShell` and `OpenSSL`.
+
+You can then validate that the certificate will load using an example such as an [aspnetcore app](https://github.com/dotnet/dotnet-docker/blob/master/samples/run-aspnetcore-https-development.md) hosted in a container.
```
[the remark this answers is not in the document]
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+```
+
+### Prepare .net 5 Sample App
+
+For this guide, the [sample aspnetapp](https://hub.docker.com/_/microsoft-dotnet-samples) should be checked for .net 5.
+
+Check sample app [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/master/samples/aspnetapp/Dockerfile) is using .net 5.
+
+Depending on the host os, the aspnet runtime may need to be updated.
+> For example, changing from `mcr.microsoft.com/dotnet/aspnet:5.0-nanoservercore-2009 AS runtime` to `mcr.microsoft.com/dotnet/aspnet:5.0-windowsservercore-ltsc2019 AS runtime` in the Dockerfile will help with targeting the appropriate Windows runtime.
```
[the remark this answers is not in the document]
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+
+Get the sample app.
+
+```console
+git clone https://github.com/dotnet/dotnet-docker/
+```
+
+Navigate to the repository locally and open up the workspace in an editor.
+
+> If you're looking to use dotnet publish parameters to *trim* the deployment, you should make sure that the appropriate dependencies are included for supporting SSL certificates.
```
[the remark this answers is not in the document]
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md:43]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 15]
```diff
@@ -0,0 +1,334 @@
...
+
+  <PropertyGroup>
+    <TargetFramework>net5.0</TargetFramework>
+    <!--Other Properties-->
+  </PropertyGroup>
+
+</Project>
+```
+
+> If you're looking to use dotnet publish parameters to *trim* the deployment, you should make sure that the appropriate dependencies are included for supporting SSL certificates.
```
[the remark this answers is not in the document]
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md:145]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+### Prepare .net 5 Sample App
+
+For this guide, the [sample aspnetapp](https://hub.docker.com/_/microsoft-dotnet-samples) should be checked for .net 5.
+
+Check sample app [Dockerfile](https://github.com/dotnet/dotnet-docker/blob/master/samples/aspnetapp/Dockerfile) is using .net 5.
+
+Depending on the host os, the aspnet runtime may need to be updated.
+> For example, changing from `mcr.microsoft.com/dotnet/aspnet:5.0-nanoservercore-2009 AS runtime` to `mcr.microsoft.com/dotnet/aspnet:5.0-windowsservercore-ltsc2019 AS runtime` in the Dockerfile will help with targeting the appropriate Windows runtime.
+
+For example, this will help with testing the certificates on `Windows`:
```
[the remark this answers is not in the document]
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+# final stage/image
+# Uses the 2009 release; 2004, 1909, and 1809 are other choices
+FROM mcr.microsoft.com/dotnet/aspnet:5.0-windowsservercore-ltsc2019 AS runtime
+WORKDIR /app
+COPY --from=build /app ./
+ENTRYPOINT ["aspnetapp"]
+
+```
+
+If we're testing the certificates on `Linux`, you can use the existing Dockerfile.
```
[the remark this answers is not in the document]
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+
+Build the container for testing locally.
+
+```console
+docker build -t aspnetapp:my-sample -f Dockerfile .
+```
+
+## Create a self-signed certificate with dotnet dev-certs
+
+You can use `dotnet dev-certs` to work with self-signed certificates. This example uses a `PowerShell` console.
```
[the remark this answers is not in the document]
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 19]
```diff
@@ -0,0 +1,334 @@
...
+
+> Note: The certificate name, in this case *aspnetapp*.pfx must match the project assembly name. `crypticpassword` is used as a stand-in for a password of your own choosing. If console returns "A valid HTTPS certificate is already present.", a trusted certificate already exists in your store. It can be exported using MMC Console.
+
+Configure application secrets, for the certificate:
+
+```console
+dotnet user-secrets -p aspnetapp\aspnetapp.csproj set "Kestrel:Certificates:Development:Password" "crypticpassword"
+```
+
+> Note: The password must match the password used for the certificate.
```
[the remark this answers is not in the document]
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md:178]: Looks good in preview version, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 20]
```diff
@@ -0,0 +1,334 @@
...
+If the secrets and certificates are not in use, be sure to clean them up.
+
+```console
+dotnet user-secrets remove "Kestrel:Certificates:Development:Password" -p aspnetapp\aspnetapp.csproj
+dotnet dev-certs https --clean
+```
+
+## Create a self-signed certificate with PowerShell
+
+You can use `PowerShell` to generate self-signed certificates. The [PKI Client](https://docs.microsoft.com/powershell/module/pkiclient/new-selfsignedcertificate?view=win10-ps&preserver-view=true) can be used to generate a self-signed certificate.
```
[the remark this answers is not in the document]
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 21]
```diff
@@ -0,0 +1,334 @@
...
+```powershell
+$certKeyPath = "c:\certs\contoso.com.pfx"
+$password = ConvertTo-SecureString 'password' -AsPlainText -Force
+$cert | Export-PfxCertificate -FilePath $certKeyPath -Password $password
+$rootCert = $(Import-PfxCertificate -FilePath $certKeyPath -CertStoreLocation 'Cert:\LocalMachine\Root' -Password $password)
+```
+
+At this point, the certificates should be viewable from an [MMC snap-in](../../framework/wcf/feature-details/how-to-view-certificates-with-the-mmc-snap-in.md).
+
+You can run the sample container in `wsl`:
```
[the remark this answers is not in the document]
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 22]
```diff
@@ -0,0 +1,334 @@
...
+
+```console
+docker run --rm -it -p 8000:80 -p 8001:443 -e ASPNETCORE_URLS="https://+;http://+" -e ASPNETCORE_HTTPS_PORT=8001 -e ASPNETCORE_ENVIRONMENT=Development -e ASPNETCORE_Kestrel__Certificates__Default__Password="password" -e ASPNETCORE_Kestrel__Certificates__Default__Path=/https/contoso.com.pfx -v /c/certs:/https/ mcr.microsoft.com/dotnet/samples:aspnetapp
+```
+
+> Note that with the volume mount the file path could be handled differently based on host.  For example, in WSL we may replace /c/certs with /mnt/c/certs.
```
[the remark this answers is not in the document]
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 23]
```diff
@@ -0,0 +1,334 @@
...
+
+If you're using the container built earlier for `Windows`, the run command would look like the following:
```
[the remark this answers is not in the document]
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 24]
```diff
@@ -0,0 +1,334 @@
...
+openssl x509 -noout -text -in $PARENT.crt
+```
+
+To get a .pfx, use the following command:
+
+```bash
+openssl pkcs12 -export -out $PARENT.pfx -inkey $PARENT.key -in $PARENT.crt
+```
+
+> The .aspnetcore 3.1 example will use `.pfx` and a password. Starting with the `.net 5` runtime, Kestrel can also take `.crt` and PEM-encoded `.key` files.
```
[the remark this answers is not in the document]
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md:292]: Added it, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 25]
```diff
@@ -0,0 +1,334 @@
...
+
+Depending on the host os, the certificate will need to be trusted. On a `Linux` host, 'trusting' the certificate is different and distro dependent.
+
+For the purposes of this guide, here's an example in `Windows` using `PowerShell`:
```
[the remark this answers is not in the document]
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md [hunk 26]
```diff
@@ -0,0 +1,334 @@
...
+
+To get a .pfx, use the following command:
+
+```bash
+openssl pkcs12 -export -out $PARENT.pfx -inkey $PARENT.key -in $PARENT.crt
+```
+
+> The .aspnetcore 3.1 example will use `.pfx` and a password. Starting with the `.net 5` runtime, Kestrel can also take `.crt` and PEM-encoded `.key` files.
+
+Depending on the host os, the certificate will need to be trusted. On a `Linux` host, 'trusting' the certificate is different and distro dependent.
```
[the remark this answers is not in the document]
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md (hunk 1 above)
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md (hunk 2 above)
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md (hunk 3 above)
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md:315]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md (hunk 4 above)
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md (hunk 5 above)
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md (hunk 6 above)
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md (hunk 7 above)
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md (hunk 8 above)
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

File: docs/core/additional-tools/self-signed-certificates-guide.md (hunk 10 above)
Comment (plooploops) [reply docs/core/additional-tools/self-signed-certificates-guide.md]: Updated, thank you!

**Review by BillWagner — APPROVED** [review]

**Comment by BillWagner** [comment]
Thanks for making the updates @plooploops I'll :shipit: now.

**Comment by plooploops** [comment]
Thanks for the approval and merge @BillWagner!

**Pull request merged**